### PR TITLE
Pybind type fixes

### PR DESCRIPTION
**Pull request by breznak**
alternative PR to #484 until that one is cleaned up 
easier to review here

### CODE REVIEW

File: bindings/py/cpp_src/bindings/engine/py_utils.hpp [hunk 1]
```diff
@@ -41,6 +41,16 @@ namespace nupic_ext {
...
 
     template<typename T> T* get_it(py::array& a) { return (T*)a.request().ptr; }
+
+    // Check that the precision in bytes matches the data size of the array
+    template<typename T> T* get_it( int precision, py::array& a)
+	{
+	if (precision != a.request().itemsize)
+		{throw std::invalid_argument("Invalid numpy array precision used.");}
+
+	return (T*)a.request().ptr;
```
Comment (breznak) [inline bindings/py/cpp_src/bindings/engine/py_utils.hpp]: nit, if we are changing this, please prefer to use `return static_cast<T*>(a.request().ptr)`

File: bindings/py/cpp_src/bindings/engine/py_utils.hpp [hunk 2]
```diff
@@ -41,6 +41,16 @@ namespace nupic_ext {
     template<typename T> T* get_end(py::array_t<T>& a) { return ((T*)a.request().ptr) + a.size(); }
 
     template<typename T> T* get_it(py::array& a) { return (T*)a.request().ptr; }
+
+    // Check that the precision in bytes matches the data size of the array
+    template<typename T> T* get_it( int precision, py::array& a)
+	{
+	if (precision != a.request().itemsize)
```
Comment (breznak) [inline bindings/py/cpp_src/bindings/engine/py_utils.hpp]: can we use only the signature `get_it<T>(py:array& a)` 
removing the precision field, and instead 
`NTA_CHECK(sizeof(T) == a.request().itemsize) << "Invalid numpy array precision used.";` //change to "Numpy type must match with C++" << string(T);   ?

Comment (breznak) [reply bindings/py/cpp_src/bindings/engine/py_utils.hpp]: and if there still is `get_it(py:array_t<T>& a);` remove that and only use the correct (precision-checking method)?

Comment (breznak) [reply bindings/py/cpp_src/bindings/engine/py_utils.hpp]: CC @fcr

Comment (breznak) [reply bindings/py/cpp_src/bindings/engine/py_utils.hpp]: also, the NTA_CHECK is optional, if the `throw ` works better for handling in python (?)

Comment (fcr) [reply bindings/py/cpp_src/bindings/engine/py_utils.hpp]: Gentlemen,
Your improvements are well beyond what I would come up with in C++ ( I am fully a untyped or inferred type language person).
I will only use the community version via Python.

File: bindings/py/cpp_src/bindings/algorithms/py_SpatialPooler.cpp [hunk 3]
```diff
@@ -186,87 +185,87 @@ using namespace sdr;
 	    ); 
 
         // setBoostFactors
-        py_SpatialPooler.def("setBoostFactors", [](SpatialPooler& self, py::array_t<Real>& x)
+        py_SpatialPooler.def("setBoostFactors", [](SpatialPooler& self, py::array& x)
```
Comment (breznak) [inline bindings/py/cpp_src/bindings/algorithms/py_SpatialPooler.cpp]: all py bindings now use py::array for the reasons @fcr found in #484

File: bindings/py/cpp_src/bindings/engine/py_utils.hpp [hunk 4]
```diff
@@ -37,21 +37,18 @@ namespace py = pybind11;
 
 namespace nupic_ext {
 
-    template<typename T> T* get_it(py::array_t<T>& a) { return (T*)a.request().ptr; }
-    template<typename T> T* get_end(py::array_t<T>& a) { return ((T*)a.request().ptr) + a.size(); }
```
Comment (breznak) [inline bindings/py/cpp_src/bindings/engine/py_utils.hpp]: replacing get_it(array_t) with get_it(array)

File: bindings/py/cpp_src/bindings/engine/py_utils.hpp [hunk 5]
```diff
@@ -37,21 +37,18 @@ namespace py = pybind11;
...
+    // Check that the precision in bytes matches the data size of the array
+    template<typename T> T* get_it(py::array& a) //TODO add explanation why array, and not array_t
+	{
+	if (sizeof(T) != a.request().itemsize)
```
Comment (breznak) [inline bindings/py/cpp_src/bindings/engine/py_utils.hpp]: implicit check for matching precision

File: bindings/py/cpp_src/bindings/engine/py_utils.hpp [hunk 6]
```diff
@@ -37,21 +37,18 @@ namespace py = pybind11;
...
+		{throw std::invalid_argument("Invalid numpy array precision used.");}
 
-    template<typename T> T* get_it(py::array& a) { return (T*)a.request().ptr; }
-    template<typename T> T* get_end(py::array& a) { return ((T*)a.request().ptr) + a.size(); }
+	return static_cast<T*>(a.request().ptr);
+	}
 
-    template<typename T> T* get_row_it(py::array_t<T>& a, int row)
```
Comment (breznak) [inline bindings/py/cpp_src/bindings/engine/py_utils.hpp]: never used, removed

File: bindings/py/cpp_src/bindings/engine/py_utils.hpp (hunk 5 above)
Comment (fcr) [reply bindings/py/cpp_src/bindings/engine/py_utils.hpp]: Nice. Much better than my version.

File: bindings/py/cpp_src/bindings/algorithms/py_SpatialPooler.cpp (hunk 3 above)
Comment (fcr) [reply bindings/py/cpp_src/bindings/algorithms/py_SpatialPooler.cpp]: I thought that they should have been changed as you did but I didn't want to make too many waves. :-)

**Comment by fcr** [comment]
@breznak In short I agree with all your fixes. Glad I was able to help.

**Comment by breznak** [comment]
>  Glad I was able to help. 

thanks, all the credit is yours, for 
- [x] enabling Anaconda python 
- [x] fixing a hidden bug in the bindings :+1: 

Thank you!

**Pull request merged**